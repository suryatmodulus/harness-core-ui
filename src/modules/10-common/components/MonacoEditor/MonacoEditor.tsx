import React, { MutableRefObject } from 'react'
import ReactMonacoEditor from 'react-monaco-editor'
import type { MonacoEditorProps } from 'react-monaco-editor'
//@ts-ignore
import { StaticServices } from 'monaco-editor/esm/vs/editor/standalone/browser/standaloneServices'
StaticServices.configurationService.get().updateValue('files.eol', '\n')
//@ts-ignore
import YamlWorker from 'worker-loader!@wings-software/monaco-yaml/lib/esm/yaml.worker'
//@ts-ignore
import EditorWorker from 'worker-loader!monaco-editor/esm/vs/editor/editor.worker'

export type ReactMonacoEditorRef =
  | ((instance: ReactMonacoEditor | null) => void)
  | MutableRefObject<ReactMonacoEditor | null>
  | null

export interface ExtendedMonacoEditorProps extends MonacoEditorProps {
  name?: string
  'data-testid'?: string
}

const MonacoEditor = (props: ExtendedMonacoEditorProps, ref: ReactMonacoEditorRef) => {
  React.useEffect(() => {
    const remeasureFonts = () => {
      //@ts-ignore
      monaco?.editor?.remeasureFonts()
    }

    // TODO: font name should be a global (for all)
    const loaded = (document as any).fonts?.check?.('1em Roboto Mono')

    if (loaded) {
      remeasureFonts()
    } else {
      ;(document as any).fonts?.ready?.then?.(remeasureFonts)
    }
  }, [])

  const editorWillMount = () => {
    monaco?.editor?.defineTheme('disable-theme', {
      base: 'vs',
      inherit: true,
      rules: [{ background: 'f3f3fa' }],
      colors: {
        'editor.background': '#f3f3fa'
      }
    })
    //@ts-ignore
    window.MonacoEnvironment = {
      getWorker(_workerId: unknown, label: string) {
        if (label === 'yaml') {
          return new YamlWorker()
        }
        return new EditorWorker()
      }
    }
  }

  const theme = props.options?.readOnly ? 'disable-theme' : 'vs'

  return <ReactMonacoEditor {...props} ref={ref} theme={theme} editorWillMount={editorWillMount} />
}

export default React.forwardRef(MonacoEditor)
