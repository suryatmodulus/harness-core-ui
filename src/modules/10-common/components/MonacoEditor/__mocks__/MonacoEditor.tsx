/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { MonacoEditorProps } from 'react-monaco-editor'

export default function MonacoEditor(
  props: MonacoEditorProps & { name: string; value?: string; 'data-testid'?: string }
) {
  return (
    <textarea
      name={props.name}
      value={props.value}
      data-testid={props['data-testid']}
      onChange={e => props.onChange?.(e.target.value, e as any)}
    />
  )
}
