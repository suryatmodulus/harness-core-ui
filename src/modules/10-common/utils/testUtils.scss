/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.browser {
  border: 1px solid #cccccc;
  border-radius: 8px;
  min-height: 300px;
  overflow: auto;

  .browser-header {
    background-color: #cccccc;
    padding: 8px 16px;
    display: flex;
  }

  .browser-path {
    display: block;
    border: none;
    line-height: 14px;
    font-size: 12px;
    color: #444444;
    border-radius: 9px;
    min-width: 100px;
    padding: 2px 8px;
    flex: 1 1 auto;
    border: 1px solid transparent;

    &:focus {
      border-color: rgb(30, 167, 253);
      outline: none;
    }
  }

  .browser-content {
    padding: 8px;
  }
}
