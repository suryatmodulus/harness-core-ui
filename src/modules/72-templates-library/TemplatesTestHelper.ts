/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { ResponseString, ResponsePageTemplateSummaryResponse } from 'services/template-ng'
import type { UseGetMockDataWithMutateAndRefetch } from '@common/utils/testUtils'

export const stageMockTemplatesInputYaml: ResponseString = {
  status: 'SUCCESS',
  data:
    'type: "Deployment"' +
    '\nspec:' +
    '\n  serviceConfig:' +
    '\n    serviceRef: "<+input>"' +
    '\n  infrastructure:' +
    '\n    infrastructureDefinition:' +
    '\n      type: "KubernetesDirect"' +
    '\n      spec:' +
    '\n        namespace: "<+input>"' +
    '\n  execution:' +
    '\n    steps:' +
    '\n    - step:' +
    '\n        identifier: "Step_1"' +
    '\n        type: "ShellScript"' +
    '\n        spec:' +
    '\n          source:' +
    '\n            type: "Inline"' +
    '\n            spec:' +
    '\n              script: "<+input>"' +
    '\n'
}

export const mockTemplatesInputYaml: ResponseString = {
  status: 'SUCCESS',
  data:
    'type: "HarnessApproval"\ntimeout: "<+input>"' +
    '\nspec:\n  approvalMessage: "<+input>"\n  approvers:\n    userGroups: "<+input>"\n    minimumCount: "<+input>"\n'
}

export const mockTemplates: ResponsePageTemplateSummaryResponse = {
  status: 'SUCCESS',
  data: {
    content: [
      {
        accountId: 'kmpySmUISimoRrJL6NL73w',
        orgIdentifier: 'default',
        projectIdentifier: 'Templateproject',
        identifier: 'manjutesttemplate',
        name: 'manju-test-template-qq-12344',
        description:
          'Flink is a versatile framework, supporting many different deployment scenarios in a mix and match fashion.',
        tags: { QAR: '', 'Internal 1': '', 'Canary 1': '', BLUE: '', 'Tag A': '' },
        yaml:
          'template:' +
          '\n    name: manju-test-template-qq-12344' +
          '\n    identifier: manjutesttemplate' +
          '\n    versionLabel: v4' +
          '\n    type: Step' +
          '\n    projectIdentifier: Templateproject' +
          '\n    orgIdentifier: default' +
          '\n    description: Flink is a versatile framework, supporting many different deployment scenarios in a mix and match fashion.' +
          '\n    tags:' +
          '\n        QAR: ""' +
          '\n        Internal 1: ""' +
          '\n        Canary 1: ""' +
          '\n        BLUE: ""' +
          '\n        Tag A: ""' +
          '\n    spec:' +
          '\n        type: HarnessApproval' +
          '\n        timeout: <+input>' +
          '\n        spec:' +
          '\n            approvalMessage: <+input>' +
          '\n            includePipelineExecutionHistory: true' +
          '\n            approvers:' +
          '\n                userGroups: <+input>' +
          '\n                minimumCount: <+input>' +
          '\n                disallowPipelineExecutor: false' +
          '\n            approverInputs:' +
          '\n                - name: "1"' +
          '\n                  defaultValue: "1"' +
          '\n                - name: ttt' +
          '\n                  defaultValue: ttt' +
          '\n        failureStrategies:' +
          '\n            - onFailure:' +
          '\n                  errors:' +
          '\n                      - Timeout' +
          '\n                  action:' +
          '\n                      type: MarkAsSuccess' +
          '\n',
        versionLabel: 'v4',
        templateEntityType: 'Step',
        childType: 'HarnessApproval',
        templateScope: 'project',
        version: 8,
        gitDetails: {},
        lastUpdatedAt: 1635626311830,
        stableTemplate: true
      }
    ],
    pageable: {
      sort: { sorted: true, unsorted: false, empty: false },
      pageSize: 20,
      pageNumber: 0,
      offset: 0,
      paged: true,
      unpaged: false
    },
    last: true,
    totalPages: 1,
    totalElements: 1,
    sort: { sorted: true, unsorted: false, empty: false },
    first: true,
    numberOfElements: 5,
    size: 20,
    number: 0,
    empty: false
  }
}

export const mockTemplatesSuccessResponse: UseGetMockDataWithMutateAndRefetch<ResponsePageTemplateSummaryResponse> = {
  loading: false,
  refetch: jest.fn(),
  mutate: jest.fn(),
  cancel: jest.fn(),
  data: mockTemplates
}
