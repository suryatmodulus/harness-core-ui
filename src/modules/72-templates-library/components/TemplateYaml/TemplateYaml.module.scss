.container {
  background-color: #f3f3fa !important;
  :global {
    .monaco-editor .rename-box {
      top: 0;
    }
  }
}
