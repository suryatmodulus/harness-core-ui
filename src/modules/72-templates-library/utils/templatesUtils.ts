import type { UseStringsReturn } from 'framework/strings'

export enum TemplateType {
  Step = 'Step',
  Stage = 'Stage',
  Pipeline = 'Pipeline',
  Service = 'Service',
  Infrastructure = 'Infrastructure',
  StepGroup = 'StepGroup',
  Execution = 'Execution'
}

export const AllTemplatesTypes = 'All'

export interface TemplateTypeOption {
  label: string
  value: string
  disabled?: boolean
}

export const getAllowedTemplateTypes = (getString: UseStringsReturn['getString']): TemplateTypeOption[] => [
  {
    label: getString('step'),
    value: TemplateType.Step
  },
  {
    label: getString('templatesLibrary.stageTemplate'),
    value: TemplateType.Stage
  },
  {
    label: getString('common.pipeline'),
    value: TemplateType.Pipeline,
    disabled: true
  },
  {
    label: getString('service'),
    value: TemplateType.Service,
    disabled: true
  },
  {
    label: getString('infrastructureText'),
    value: TemplateType.Infrastructure,
    disabled: true
  },
  {
    label: getString('stepGroup'),
    value: TemplateType.StepGroup,
    disabled: true
  },
  {
    label: getString('executionText'),
    value: TemplateType.Execution,
    disabled: true
  }
]
