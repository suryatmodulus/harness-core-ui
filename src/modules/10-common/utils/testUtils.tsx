/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { UseGetProps, UseGetReturn, RestfulProvider } from 'restful-react'
import { queryByAttribute } from '@testing-library/react'
import { compile } from 'path-to-regexp'
import { createMemoryHistory } from 'history'
import { Router, Route, Switch, useLocation, useHistory } from 'react-router-dom'
import { ModalProvider } from '@wings-software/uicore'
import qs from 'qs'

import { enableMapSet } from 'immer'
import { AppStoreContext, AppStoreContextProps } from 'framework/AppStore/AppStoreContext'
import { LicenseStoreContext, LicenseStoreContextProps } from 'framework/LicenseStore/LicenseStoreContext'
import { LICENSE_STATE_VALUES } from 'framework/LicenseStore/licenseStoreUtil'
import { withAccountId, accountPathProps } from '@common/utils/routeUtils'
import type { Project } from 'services/cd-ng'
import { StringsContext } from 'framework/strings'

import { FeaturesContext, FeaturesContextProps } from 'framework/featureStore/FeaturesContext'
import type { FeatureDetail, FeatureMetaData } from 'framework/featureStore/featureStoreUtil'
import type { FeatureIdentifier } from 'framework/featureStore/FeatureIdentifier'
import './testUtils.scss'
import { PermissionsContext, PermissionsContextProps } from 'framework/rbac/PermissionsContext'

export type UseGetMockData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Required<
  UseGetProps<TData, TError, TQueryParams, TPathParams>
>['mock']

export interface UseGetMockDataWithMutateAndRefetch<T> extends UseGetMockData<T> {
  mutate: () => Record<string, unknown>
  refetch: () => Record<string, unknown>
  cancel?: () => Record<string, unknown>
}

export interface UseMutateMockData<TData, TRequestBody = unknown> {
  loading?: boolean
  mutate?: (data?: TRequestBody) => Promise<TData>
}

export type UseGetReturnData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Omit<
  UseGetReturn<TData, TError, TQueryParams, TPathParams>,
  'absolutePath' | 'cancel' | 'response'
>

export const findDialogContainer = (): HTMLElement | null => document.querySelector('.bp3-dialog')
export const findPopoverContainer = (): HTMLElement | null => document.querySelector('.bp3-popover-content')

export interface TestWrapperProps {
  path?: string
  pathParams?: Record<string, string | number>
  queryParams?: Record<string, unknown>
  defaultAppStoreValues?: Partial<AppStoreContextProps>
  defaultLicenseStoreValues?: Partial<LicenseStoreContextProps>
  defaultPermissionValues?: Partial<PermissionsContextProps>
  defaultFeaturesValues?: Partial<FeaturesContextProps>
  projects?: Project[]
  enableBrowserView?: boolean
  stringsData?: Record<string, string>
  getString?(key: string): string
}

export const prependAccountPath = (path: string): string => withAccountId(() => path)(accountPathProps)

export const CurrentLocation = (): JSX.Element => {
  const location = useLocation()
  return (
    <div>
      <h1>Not Found</h1>
      <div data-testid="location">{`${location.pathname}${
        location.search ? `?${location.search.replace(/^\?/g, '')}` : ''
      }`}</div>
    </div>
  )
}

export interface BrowserViewProps {
  enable?: boolean
  children: React.ReactNode
}

export function BrowserView(props: BrowserViewProps): React.ReactElement {
  const { enable, children } = props
  const location = useLocation()
  const history = useHistory()

  if (!enable) {
    return <>{children}</>
  }

  function handlePathChange(e: React.ChangeEvent<HTMLInputElement>): void {
    history.replace(e.currentTarget.value)
  }

  const search = location.search ? `?${location.search.replace(/^\?/, '')}` : ''

  return (
    <div className="browser">
      <div className="browser-header">
        <input className="browser-path" value={location.pathname + search} onChange={handlePathChange} />
      </div>
      <div className="browser-content">{children}</div>
    </div>
  )
}

export const TestWrapper: React.FC<TestWrapperProps> = props => {
  enableMapSet()
  const {
    path = '/',
    pathParams = {},
    defaultAppStoreValues,
    queryParams = {},
    defaultLicenseStoreValues,
    defaultPermissionValues,
    defaultFeaturesValues,
    stringsData = {},
    getString = (key: string) => key
  } = props

  const search = qs.stringify(queryParams, { addQueryPrefix: true })
  const routePath = compile(path)(pathParams) + search

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const history = React.useMemo(() => createMemoryHistory({ initialEntries: [routePath] }), [])

  const defaultReturn = {
    enabled: true
  }

  /** TODO: Try fixing this later. This is causing some tests to fail */
  // React.useEffect(() => {
  //   history.replace(compile(path)(pathParams) + qs.stringify(queryParams, { addQueryPrefix: true }))
  //   // eslint-disable-next-line react-hooks/exhaustive-deps
  // }, [path, pathParams, queryParams])

  return (
    <StringsContext.Provider value={{ data: stringsData as any, getString }}>
      <AppStoreContext.Provider
        value={{
          featureFlags: {
            FEATURE_ENFORCEMENT_ENABLED: true
          },
          updateAppStore: () => void 0,
          currentUserInfo: {},
          ...defaultAppStoreValues
        }}
      >
        <LicenseStoreContext.Provider
          value={{
            versionMap: {},
            licenseInformation: {},
            CI_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
            FF_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
            CCM_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
            CD_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
            updateLicenseStore: () => void 0,
            ...defaultLicenseStoreValues
          }}
        >
          <PermissionsContext.Provider
            value={{
              permissions: new Map<string, boolean>(),
              requestPermission: () => void 0,
              checkPermission: () => true,
              cancelRequest: () => void 0,
              ...defaultPermissionValues
            }}
          >
            <FeaturesContext.Provider
              value={{
                features: new Map<FeatureIdentifier, FeatureDetail>(),
                featureMap: new Map<FeatureIdentifier, FeatureMetaData>(),
                getEdition: () => {
                  return undefined
                },
                requestFeatures: () => void 0,
                requestLimitFeature: () => void 0,
                checkFeature: () => {
                  return defaultReturn
                },
                checkLimitFeature: () => {
                  return defaultReturn
                },
                getRestrictionType: () => {
                  return undefined
                },
                ...defaultFeaturesValues
              }}
            >
              <Router history={history}>
                <ModalProvider>
                  <RestfulProvider base="/">
                    <BrowserView enable={props.enableBrowserView}>
                      <Switch>
                        <Route exact path={path}>
                          {props.children}
                        </Route>
                        <Route>
                          <CurrentLocation />
                        </Route>
                      </Switch>
                    </BrowserView>
                  </RestfulProvider>
                </ModalProvider>
              </Router>
            </FeaturesContext.Provider>
          </PermissionsContext.Provider>
        </LicenseStoreContext.Provider>
      </AppStoreContext.Provider>
    </StringsContext.Provider>
  )
}

export const queryByNameAttribute = (name: string, container: HTMLElement): HTMLElement | null =>
  queryByAttribute('name', container, name)
