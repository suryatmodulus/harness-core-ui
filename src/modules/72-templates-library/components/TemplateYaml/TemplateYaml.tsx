import React from 'react'
import { Color, Container, Text, Layout } from '@wings-software/uicore'
import { defaultTo } from 'lodash-es'
import MonacoEditor from '@common/components/MonacoEditor/MonacoEditor'

import { yamlStringify } from '@common/utils/YamlHelperMethods'
import css from './TemplateYaml.module.scss'

export interface TemplateYamlProps {
  templateYaml?: string
}

export const TemplateYaml: React.FC<TemplateYamlProps> = ({ templateYaml }) => {
  const [height, setHeight] = React.useState<number>()

  React.useEffect(() => {
    const stringifiedYaml = yamlStringify(templateYaml) as string
    setHeight((defaultTo(stringifiedYaml.match(/\n/g)?.length, 0) + 1) * 20)
  }, [templateYaml])

  return (
    <Container className={css.container}>
      <Layout.Vertical spacing={'medium'}>
        <Container
          padding={{ top: 'large', right: 'xxlarge', bottom: 'large', left: 'xxlarge' }}
          border={{ bottom: true }}
          background={Color.WHITE}
        >
          <Text font={{ size: 'normal', weight: 'bold' }} color={Color.GREY_800}>
            template.yaml
          </Text>
        </Container>
        <Container style={{ flexGrow: 1 }}>
          <MonacoEditor
            value={templateYaml}
            language={'yaml'}
            height={height}
            options={
              {
                fontFamily: "'Roboto Mono', monospace",
                fontSize: 13,
                minimap: {
                  enabled: false
                },
                readOnly: true,
                scrollbar: {
                  handleMouseWheel: false
                },
                overviewRulerLanes: 0,
                scrollBeyondLastLine: false
              } as any
            }
          />
        </Container>
      </Layout.Vertical>
    </Container>
  )
}
