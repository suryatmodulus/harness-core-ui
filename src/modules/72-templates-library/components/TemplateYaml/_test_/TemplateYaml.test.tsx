/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { mockTemplatesInputYaml } from '@templates-library/TemplatesTestHelper'
import MonacoEditor from '@common/components/MonacoEditor/__mocks__/MonacoEditor'
import { TemplateYaml } from '../TemplateYaml'

jest.mock('react-monaco-editor', () => ({
  MonacoDiffEditor: MonacoEditor
}))

jest.mock('@common/components/MonacoEditor/MonacoEditor', () => MonacoEditor)

describe('<TemplateYaml /> tests', () => {
  test('snapshot test', async () => {
    const { container } = render(
      <TestWrapper>
        <TemplateYaml templateYaml={mockTemplatesInputYaml.data} />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
