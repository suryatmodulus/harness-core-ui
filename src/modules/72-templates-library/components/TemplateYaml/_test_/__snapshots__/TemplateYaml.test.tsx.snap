// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateYaml /> tests snapshot test 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main container"
  >
    <div
      class="Layout--vertical Layout--layout-spacing-medium StyledProps--main"
    >
      <div
        class="StyledProps--font StyledProps--main StyledProps--padding-top-large StyledProps--padding-right-xxlarge StyledProps--padding-bottom-large StyledProps--padding-left-xxlarge StyledProps--border StyledProps--border-bottom StyledProps--background-white"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-size-normal StyledProps--font-weight-bold StyledProps--color StyledProps--color-grey800"
        >
          template.yaml
        </p>
      </div>
      <div
        class="StyledProps--font StyledProps--main"
        style="flex-grow: 1;"
      >
        <textarea>
          type: "HarnessApproval"
timeout: "&lt;+input&gt;"
spec:
  approvalMessage: "&lt;+input&gt;"
  approvers:
    userGroups: "&lt;+input&gt;"
    minimumCount: "&lt;+input&gt;"

        </textarea>
      </div>
    </div>
  </div>
</div>
`;
